### ISSUE TITLE
APIs failing after connecteddrive backend update

### ISSUE THREAD

**Issue by davecarlson** [bimmerconnected/bimmer_connected#434]
### Describe the issue

The status endpoints are not working. 
I know that a new update to their server backend was pushed today.

DEBUG:urllib3.connectionpool:https://cocoapi.bmwgroup.com:443 "GET /eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1652892002377&tireGuardMode=ENABLED HTTP/1.1" 500 111
ERROR:bimmer_connected.account:MyBMW API error '500 Server Error: Internal Server Error for url: https://cocoapi.bmwgroup.com/eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1652892002377&tireGuardMode=ENABLED': { "statusCode": 500, "message": "Internal server error", "activityId": "10ad07a2-9688-4bb6-9b64-6a5da7f3c0cd" }
Traceback (most recent call last):
  File "/opt/homebrew/bin/bimmerconnected", line 8, in <module>
    sys.exit(main())
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/cli.py", line 282, in main
    args.func(args)
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/cli.py", line 101, in get_status
    account = ConnectedDriveAccount(args.username, args.password, get_region_from_name(args.region))
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/account.py", line 88, in __init__
    self._get_vehicles()
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/account.py", line 386, in _get_vehicles
    response = self.send_request(
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/account.py", line 313, in send_request
    raise ex
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/account.py", line 306, in send_request
    response.raise_for_status()
  File "/opt/homebrew/lib/python3.9/site-packages/requests/models.py", line 960, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: https://cocoapi.bmwgroup.com/eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1652892002377&tireGuardMode=ENABLED

### Expected behavior

It should bring back car data

### Which Home Assistant version are you using?

not

### What is your region?

Rest of world

### MyBMW website

- [X] I can still successfully login to the BMW MyBMW website and the car status is available there.
- [X] I have MyBMW enabled for my vehicle.

### Number of cars

- [ ] I have 2 or more cars linked to the MyBMW account.
- [ ] I have a Mini vehicle linked to my account.

### Anything in the logs that might be useful for us?

```txt
$ bimmerconnected status CREDS
DEBUG:bimmer_connected.account:Getting vehicle list
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): cocoapi.bmwgroup.com:443
eDEBUG:urllib3.connectionpool:https://cocoapi.bmwgroup.com:443 "GET /eadrax-ucs/v1/presentation/oauth/config HTTP/1.1" 200 580
DEBUG:bimmer_connected.account:Authenticating against GCDM with MyBMW flow.
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): customer.bmwgroup.com:443
DEBUG:urllib3.connectionpool:https://customer.bmwgroup.com:443 "POST /gcdm/oauth/authenticate HTTP/1.1" 200 411
DEBUG:urllib3.connectionpool:https://customer.bmwgroup.com:443 "POST /gcdm/oauth/authenticate HTTP/1.1" 302 0
eDEBUG:urllib3.connectionpool:https://customer.bmwgroup.com:443 "POST /gcdm/oauth/token HTTP/1.1" 200 1962
DEBUG:bimmer_connected.account:got new token HIDDEN with expiration date 2022-05-18 17:39:51.283206
DEBUG:bimmer_connected.account:Request to: https://cocoapi.bmwgroup.com/eadrax-vcs/v1/vehicles
DEBUG:bimmer_connected.account:Old token is still valid. Not getting a new one.
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): cocoapi.bmwgroup.com:443
DEBUG:urllib3.connectionpool:https://cocoapi.bmwgroup.com:443 "GET /eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1652892002377&tireGuardMode=ENABLED HTTP/1.1" 500 111
ERROR:bimmer_connected.account:MyBMW API error '500 Server Error: Internal Server Error for url: https://cocoapi.bmwgroup.com/eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1652892002377&tireGuardMode=ENABLED': { "statusCode": 500, "message": "Internal server error", "activityId": "10ad07a2-9688-4bb6-9b64-6a5da7f3c0cd" }
Traceback (most recent call last):
  File "/opt/homebrew/bin/bimmerconnected", line 8, in <module>
    sys.exit(main())
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/cli.py", line 282, in main
    args.func(args)
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/cli.py", line 101, in get_status
    account = ConnectedDriveAccount(args.username, args.password, get_region_from_name(args.region))
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/account.py", line 88, in __init__
    self._get_vehicles()
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/account.py", line 386, in _get_vehicles
    response = self.send_request(
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/account.py", line 313, in send_request
    raise ex
  File "/opt/homebrew/lib/python3.9/site-packages/bimmer_connected/account.py", line 306, in send_request
    response.raise_for_status()
  File "/opt/homebrew/lib/python3.9/site-packages/requests/models.py", line 960, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: https://cocoapi.bmwgroup.com/eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1652892002377&tireGuardMode=ENABLED
```

**Comment by rikroe** [comment]
Does it work for you again? I had lots of issues in my HA log, but also my private internet connection was broken for almost a week.

**Comment by nottledim** [comment]
It's gone wrong for me too. BMW i3 rest-of-world.  Both bimmerconnected cli and HA fail similarly.

The main error seems to be http return 429 but it's then followed by too many retries error 401.  I don't think my error message adds anything to info from original poster.
```
DEBUG:bimmer_connected.account:Getting vehicle list
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): cocoapi.bmwgroup.com:443
DEBUG:urllib3.connectionpool:https://cocoapi.bmwgroup.com:443 "GET /eadrax-ucs/v1/presentation/oauth/config HTTP/1.1" 429 84
DEBUG:urllib3.util.retry:Incremented Retry for (url='/eadrax-ucs/v1/presentation/oauth/config'): Retry(total=2, connect=3, read=3, redirect=None, status=None)
DEBUG:urllib3.connectionpool:Retry: /eadrax-ucs/v1/presentation/oauth/config
DEBUG:urllib3.connectionpool:https://cocoapi.bmwgroup.com:443 "GET /eadrax-ucs/v1/presentation/oauth/config HTTP/1.1" 429 83
DEBUG:urllib3.util.retry:Incremented Retry for (url='/eadrax-ucs/v1/presentation/oauth/config'): Retry(total=1, connect=3, read=3, redirect=None, status=None)
DEBUG:urllib3.connectionpool:Retry: /eadrax-ucs/v1/presentation/oauth/config
DEBUG:urllib3.connectionpool:https://cocoapi.bmwgroup.com:443 "GET /eadrax-ucs/v1/presentation/oauth/config HTTP/1.1" 200 580
DEBUG:bimmer_connected.account:Authenticating against GCDM with MyBMW flow.
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): customer.bmwgroup.com:443
DEBUG:urllib3.connectionpool:https://customer.bmwgroup.com:443 "POST /gcdm/oauth/authenticate HTTP/1.1" 401 190
DEBUG:urllib3.util.retry:Incremented Retry for (url='/gcdm/oauth/authenticate'): Retry(total=2, connect=3, read=3, redirect=None, status=None)
DEBUG:urllib3.connectionpool:Retry: /gcdm/oauth/authenticate
DEBUG:urllib3.connectionpool:https://customer.bmwgroup.com:443 "POST /gcdm/oauth/authenticate HTTP/1.1" 401 190
DEBUG:urllib3.util.retry:Incremented Retry for (url='/gcdm/oauth/authenticate'): Retry(total=1, connect=3, read=3, redirect=None, status=None)
DEBUG:urllib3.connectionpool:Retry: /gcdm/oauth/authenticate
DEBUG:urllib3.connectionpool:https://customer.bmwgroup.com:443 "POST /gcdm/oauth/authenticate HTTP/1.1" 401 190
DEBUG:urllib3.util.retry:Incremented Retry for (url='/gcdm/oauth/authenticate'): Retry(total=0, connect=3, read=3, redirect=None, status=None)
DEBUG:urllib3.connectionpool:Retry: /gcdm/oauth/authenticate
DEBUG:urllib3.connectionpool:https://customer.bmwgroup.com:443 "POST /gcdm/oauth/authenticate HTTP/1.1" 401 190
ERROR:bimmer_connected.account:HTTPSConnectionPool(host='customer.bmwgroup.com', port=443): Max retries exceeded with url: /gcdm/oauth/authenticate (Caused by ResponseError('too many 401 error responses'))
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/requests/adapters.py", line 440, in send
    resp = conn.urlopen(
  File "/usr/lib/python3.10/site-packages/urllib3/connectionpool.py", line 876, in urlopen
    return self.urlopen(
  File "/usr/lib/python3.10/site-packages/urllib3/connectionpool.py", line 876, in urlopen
    return self.urlopen(
  File "/usr/lib/python3.10/site-packages/urllib3/connectionpool.py", line 876, in urlopen
    return self.urlopen(
  File "/usr/lib/python3.10/site-packages/urllib3/connectionpool.py", line 866, in urlopen
    retries = retries.increment(method, url, response=response, _pool=self)
  File "/usr/lib/python3.10/site-packages/urllib3/util/retry.py", line 592, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='customer.bmwgroup.com', port=443): Max retries exceeded with url: /gcdm/oauth/authenticate (Caused by ResponseError('too many 401 error responses'))
``

**Comment by rikroe** [comment]
If you are using the custom component, please update to the latest version which includes some major rework in terms of how authentication (and authentication issues) are handled.

Cannot say much about the 500 Internal Server Error, except that there is apparently a new `/eadrax-vcs/v2/vehicles` API and you then have to get the state of each vehicle separately (using `/eadrax-vcs/v2/vehicles/{VIN}/state`. However this is currently not implemented by the library.

**Comment by nottledim** [comment]
Sometimes it works, sometimes it doesn't.  It was working yesterday afternoon. This is in HA log now:
```
Logger: bimmer_connected.account
Source: /usr/local/lib/python3.9/site-packages/bimmer_connected/account.py:312
First occurred: 1:33:15 PM (1 occurrences)
Last logged: 1:33:15 PM

MyBMW API error '500 Server Error: Internal Server Error
etc
```
And 15 mins later 13:45, it's working again :-)

That's with latest sw (calls itself 0.8.12).

What's causing the rapid retries that the server is objecting to? 
```
Max retries exceeded with url: /gcdm/oauth/authenticate (Caused by ResponseError('too many 401 error responses'))
```
Is that something bimmer is doing (and could be stopped from doing)?

**Comment by rikroe** [comment]
Not sure why the API is so flaky at the moment - sometimes it also doesn't work if you're trying to do a fresh login (using username + password) from the app.

In our beta versions for 0.9.0 the authentication has been rewritten to not login with username/password but with an refresh_token provided by the API (same as the app doesn, username/password is only required on first login).

Unfortunately, you cannot just install the latest version to your current HA instance, it would break the component.
You can use the [custom component](https://github.com/bimmerconnected/ha_custom_component) until https://github.com/home-assistant/core/pull/71827 is merged (not sure if it will make it into `2022.6` or we have to wait for `2022.7`).

**Comment by nottledim** [comment]
Oh well, custom component works OK (once I'd realised that the battery level sensor had changed name and I spelt "battery" with a 'y'.  :-D ).  Thanks

Incidentally custom component has the advantage that it uses the python version that HA is using. I use HA  docker container which runs python 3.9 whereas my system runs python 3.10.

**Comment by rikroe** [comment]
If using the integrated component, you should be using 3.9 as well?
Python 3.10 should be fine, Python 3.6 might create some issues in the future. Nevertheless, at least all of our tests are executed against Python 3.6 to 3.10 (e.g. https://github.com/bimmerconnected/bimmer_connected/actions/runs/2366016581).

**Comment by rikroe** [comment]
Closing this as the problem should be solved with the latest custom component. The changes there will be included in the HA 2022.6 release, so it's possible to revert back once released.
